diff -apruN php-5.3.29.orig/ext/gd/gd.c php-5.3.29/ext/gd/gd.c
--- php-5.3.29.orig/ext/gd/gd.c	2016-08-24 17:50:03.000000000 -0400
+++ php-5.3.29/ext/gd/gd.c	2016-08-24 17:51:16.339121306 -0400
@@ -1673,11 +1673,11 @@ PHP_FUNCTION(imagetruecolortopalette)
 
 	ZEND_FETCH_RESOURCE(im, gdImagePtr, &IM, -1, "Image", le_gd);
 
-	if (ncolors <= 0) {
-		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Number of colors has to be greater than zero");
+	if (ncolors <= 0 || ncolors > INT_MAX) {
+		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Number of colors has to be greater than zero and no more than %d", INT_MAX);
 		RETURN_FALSE;
 	}
-	gdImageTrueColorToPalette(im, dither, ncolors);
+	gdImageTrueColorToPalette(im, dither, (int)ncolors);
 
 	RETURN_TRUE;
 }
diff -apruN php-5.3.29.orig/ext/gd/tests/bug72697.phpt php-5.3.29/ext/gd/tests/bug72697.phpt
--- php-5.3.29.orig/ext/gd/tests/bug72697.phpt	1969-12-31 19:00:00.000000000 -0500
+++ php-5.3.29/ext/gd/tests/bug72697.phpt	2016-08-24 17:51:16.340121433 -0400
@@ -0,0 +1,17 @@
+--TEST--
+Bug #72697: select_colors write out-of-bounds
+--SKIPIF--
+<?php 
+if (!function_exists("imagecreatetruecolor")) die("skip");
+if (PHP_INT_MAX !== 9223372036854775807) die("skip for 64-bit long systems only");
+?>
+--FILE--
+<?php
+
+$img=imagecreatetruecolor(10, 10);
+imagetruecolortopalette($img, false, PHP_INT_MAX / 8);
+?>
+DONE
+--EXPECTF--
+Warning: imagetruecolortopalette(): Number of colors has to be greater than zero and no more than 2147483647 in %sbug72697.php on line %d
+DONE
\ No newline at end of file
diff -apruN php-5.3.29.orig/ext/gd/tests/imagetruecolortopalette_error3.phpt php-5.3.29/ext/gd/tests/imagetruecolortopalette_error3.phpt
--- php-5.3.29.orig/ext/gd/tests/imagetruecolortopalette_error3.phpt	2014-08-13 15:22:50.000000000 -0400
+++ php-5.3.29/ext/gd/tests/imagetruecolortopalette_error3.phpt	2016-08-25 00:39:52.362981872 -0400
@@ -25,4 +25,4 @@ Warning: imagetruecolortopalette() expec
 
 Warning: imagetruecolortopalette() expects parameter 3 to be long, array given in %s on line %d
 
-Warning: imagetruecolortopalette(): Number of colors has to be greater than zero in %s on line %d
\ No newline at end of file
+Warning: imagetruecolortopalette(): Number of colors has to be greater than zero and no more than %d in %s on line %d
diff -apruN php-5.3.29.orig/ext/gd/tests/imagetruecolortopalette_error4.phpt php-5.3.29/ext/gd/tests/imagetruecolortopalette_error4.phpt
--- php-5.3.29.orig/ext/gd/tests/imagetruecolortopalette_error4.phpt	2014-08-13 15:22:50.000000000 -0400
+++ php-5.3.29/ext/gd/tests/imagetruecolortopalette_error4.phpt	2016-08-25 00:39:52.365982253 -0400
@@ -16,6 +16,6 @@ imagetruecolortopalette($image, true, -1
 
 ?>
 --EXPECTF--
-Warning: imagetruecolortopalette(): Number of colors has to be greater than zero in %s on line %d
+Warning: imagetruecolortopalette(): Number of colors has to be greater than zero and no more than %d in %s on line %d
 
-Warning: imagetruecolortopalette(): Number of colors has to be greater than zero in %s on line %d
\ No newline at end of file
+Warning: imagetruecolortopalette(): Number of colors has to be greater than zero and no more than %d in %s on line %d
