diff -apruN php-5.3.29.orig/ext/phar/phar_object.c php-5.3.29/ext/phar/phar_object.c
--- php-5.3.29.orig/ext/phar/phar_object.c	2016-12-12 05:50:12.567830157 -0500
+++ php-5.3.29/ext/phar/phar_object.c	2016-12-12 06:10:46.321633921 -0500
@@ -604,6 +604,7 @@ PHP_METHOD(Phar, webPhar)
 	}
 
 	if ((strlen(sapi_module.name) == sizeof("cgi-fcgi")-1 && !strncmp(sapi_module.name, "cgi-fcgi", sizeof("cgi-fcgi")-1))
+		|| (strlen(sapi_module.name) == sizeof("fpm-fcgi")-1 && !strncmp(sapi_module.name, "fpm-fcgi", sizeof("fpm-fcgi")-1))
 		|| (strlen(sapi_module.name) == sizeof("cgi")-1 && !strncmp(sapi_module.name, "cgi", sizeof("cgi")-1))) {
 
 		if (PG(http_globals)[TRACK_VARS_SERVER]) {
