diff -apruN php-5.3.29.orig/ext/bz2/bz2.c php-5.3.29/ext/bz2/bz2.c
--- php-5.3.29.orig/ext/bz2/bz2.c	2016-11-12 14:26:58.329664985 -0500
+++ php-5.3.29/ext/bz2/bz2.c	2016-11-12 14:27:00.827982509 -0500
@@ -515,7 +515,7 @@ static PHP_FUNCTION(bzcompress)
 	dest_len   = (unsigned int) (source_len + (0.01 * source_len) + 600);
 
 	/* Allocate the destination buffer */
-	dest = emalloc(dest_len + 1);
+	dest = safe_emalloc(dest_len, 1, 1);
 
 	/* Handle the optional arguments */
 	if (argc > 1) {
@@ -535,7 +535,7 @@ static PHP_FUNCTION(bzcompress)
 		   so we erealloc() the buffer to the proper size */
 		dest = erealloc(dest, dest_len + 1);
 		dest[dest_len] = 0;
-		RETURN_STRINGL(dest, dest_len, 0);
+		RETURN_STRINGL_CHECK(dest, dest_len, 0);
 	}
 }
 /* }}} */
diff -apruN php-5.3.29.orig/Zend/zend_API.h php-5.3.29/Zend/zend_API.h
--- php-5.3.29.orig/Zend/zend_API.h	2016-11-14 00:53:13.000000000 -0500
+++ php-5.3.29/Zend/zend_API.h	2016-11-14 01:14:04.668288683 -0500
@@ -623,6 +623,7 @@ END_EXTERN_C()
 	}                                       \
 	RETVAL_STRINGL((s), (int)__len, (dup)); \
 } while (0)
+#define RETURN_STRINGL_CHECK(s, len, dup) { RETVAL_STRINGL_CHECK(s, len, dup); return; }
 
 #define SET_VAR_STRING(n, v) {																				\
 								{																			\
