diff -apruN php-5.3.29.orig/ext/standard/uuencode.c php-5.3.29/ext/standard/uuencode.c
--- php-5.3.29.orig/ext/standard/uuencode.c	2014-08-13 15:22:50.000000000 -0400
+++ php-5.3.29/ext/standard/uuencode.c	2016-08-25 04:04:23.169538286 -0400
@@ -153,7 +153,7 @@ PHPAPI int php_uudecode(char *src, int s
 		while (s < ee) {
 			if(s+4 > e) {
 				goto err;
-			} 
+			}
 			*p++ = PHP_UU_DEC(*s) << 2 | PHP_UU_DEC(*(s + 1)) >> 4;
 			*p++ = PHP_UU_DEC(*(s + 1)) << 4 | PHP_UU_DEC(*(s + 2)) >> 2;
 			*p++ = PHP_UU_DEC(*(s + 2)) << 6 | PHP_UU_DEC(*(s + 3));
@@ -188,7 +188,7 @@ err:
 }
 /* }}} */
 
-/* {{{ proto string convert_uuencode(string data) 
+/* {{{ proto string convert_uuencode(string data)
    uuencode a string */
 PHP_FUNCTION(convert_uuencode)
 {
@@ -200,6 +200,11 @@ PHP_FUNCTION(convert_uuencode)
 	}
 
 	dst_len = php_uuencode(src, src_len, &dst);
+	if (dst_len < 0) {
+		php_error_docref(NULL TSRMLS_CC, E_WARNING, "String too long, max length is %d", INT_MAX);
+		efree(dst);
+		RETURN_FALSE;
+	}
 
 	RETURN_STRINGL(dst, dst_len, 0);
 }
