diff -apruN php-5.3.29.orig/ext/openssl/openssl.c php-5.3.29/ext/openssl/openssl.c
--- php-5.3.29.orig/ext/openssl/openssl.c	2016-10-20 07:57:57.807975112 -0400
+++ php-5.3.29/ext/openssl/openssl.c	2016-10-20 08:32:15.054440559 -0400
@@ -4799,7 +4799,7 @@ PHP_FUNCTION(openssl_encrypt)
 	free_iv = php_openssl_validate_iv(&iv, &iv_len, max_iv_len TSRMLS_CC);
 
 	outlen = data_len + EVP_CIPHER_block_size(cipher_type);
-	outbuf = emalloc(outlen + 1);
+	outbuf = safe_emalloc(outlen, 1, 1);
 
 	EVP_EncryptInit(&cipher_ctx, cipher_type, NULL, NULL);
 	if (password_len > keylen) {
@@ -4814,14 +4814,18 @@ PHP_FUNCTION(openssl_encrypt)
 		outlen += i;
 		if (raw_output) {
 			outbuf[outlen] = '\0';
-			RETVAL_STRINGL((char *)outbuf, outlen, 0);
+			RETVAL_STRINGL_CHECK((char *)outbuf, outlen, 0);
 		} else {
 			int base64_str_len;
 			char *base64_str;
 
 			base64_str = (char*)php_base64_encode(outbuf, outlen, &base64_str_len);
 			efree(outbuf);
-			RETVAL_STRINGL(base64_str, base64_str_len, 0);
+			if (!base64_str) {
+				RETVAL_FALSE;
+			} else {
+				RETVAL_STRINGL(base64_str, base64_str_len, 0);
+			}
 		}
 	} else {
 		efree(outbuf);
