diff -apruN php-5.3.29.orig/ext/phar/phar.c php-5.3.29/ext/phar/phar.c
--- php-5.3.29.orig/ext/phar/phar.c	2017-01-25 10:11:29.074765665 -0500
+++ php-5.3.29/ext/phar/phar.c	2017-01-25 10:11:32.016139498 -0500
@@ -1054,7 +1054,7 @@ static int phar_parse_pharfile(php_strea
 	entry.is_persistent = mydata->is_persistent;
 
 	for (manifest_index = 0; manifest_index < manifest_count; ++manifest_index) {
-		if (buffer + 24 > endbuffer) {
+		if (buffer + 28 > endbuffer) {
 			MAPPHAR_FAIL("internal corruption of phar \"%s\" (truncated manifest entry)")
 		}
 
@@ -1068,7 +1068,7 @@ static int phar_parse_pharfile(php_strea
 			entry.manifest_pos = manifest_index;
 		}
 
-		if (entry.filename_len > endbuffer - buffer - 20) {
+		if (entry.filename_len > endbuffer - buffer - 24) {
 			MAPPHAR_FAIL("internal corruption of phar \"%s\" (truncated manifest entry)");
 		}
 
