diff -apruN php-5.3.29.orig/ext/standard/iptc.c php-5.3.29/ext/standard/iptc.c
--- php-5.3.29.orig/ext/standard/iptc.c	2014-08-13 15:22:50.000000000 -0400
+++ php-5.3.29/ext/standard/iptc.c	2016-02-05 14:37:52.794039850 -0500
@@ -38,6 +38,15 @@
 
 #include <sys/stat.h>
 
+#ifdef PHP_WIN32
+# include "win32/php_stdint.h"
+#else
+# if HAVE_INTTYPES_H
+#  include <inttypes.h>
+# elif HAVE_STDINT_H
+#  include <stdint.h>
+# endif
+#endif
 
 /* some defines for the different JPEG block types */
 #define M_SOF0  0xC0            /* Start Of Frame N */
@@ -76,7 +85,7 @@
 /* {{{ php_iptc_put1
  */
 static int php_iptc_put1(FILE *fp, int spool, unsigned char c, unsigned char **spoolbuf TSRMLS_DC)
-{ 
+{
 	if (spool > 0)
 		PUTC(c);
 
@@ -89,7 +98,7 @@ static int php_iptc_put1(FILE *fp, int s
 /* {{{ php_iptc_get1
  */
 static int php_iptc_get1(FILE *fp, int spool, unsigned char **spoolbuf TSRMLS_DC)
-{ 	
+{
 	int c;
 	char cc;
 
@@ -121,7 +130,7 @@ static int php_iptc_read_remaining(FILE
 /* {{{ php_iptc_skip_variable
  */
 static int php_iptc_skip_variable(FILE *fp, int spool, unsigned char **spoolbuf TSRMLS_DC)
-{ 
+{
 	unsigned int  length;
 	int c1, c2;
 
@@ -202,6 +211,11 @@ PHP_FUNCTION(iptcembed)
 		RETURN_FALSE;
 	}
 
+	if ((size_t)iptcdata_len >= SIZE_MAX - sizeof(psheader) - 1025) {
+ 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "IPTC data too large");
+ 		RETURN_FALSE;
+ 	}
+
 	if ((fp = VCWD_FOPEN(jpeg_file, "rb")) == 0) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unable to open %s", jpeg_file);
 		RETURN_FALSE;
@@ -210,9 +224,9 @@ PHP_FUNCTION(iptcembed)
 	if (spool < 2) {
 		fstat(fileno(fp), &sb);
 
-		poi = spoolbuf = safe_emalloc(1, iptcdata_len + sizeof(psheader) + sb.st_size + 1024, 1);
+		poi = spoolbuf = safe_emalloc(1, (size_t)iptcdata_len + sizeof(psheader) + 1024 + 1, sb.st_size);
 		memset(poi, 0, iptcdata_len + sizeof(psheader) + sb.st_size + 1024 + 1);
-	} 
+	}
 
 	if (php_iptc_get1(fp, spool, poi?&poi:0 TSRMLS_CC) != 0xFF) {
 		fclose(fp);
