diff -apruN php-5.3.29.orig/ext/dom/document.c php-5.3.29/ext/dom/document.c
--- php-5.3.29.orig/ext/dom/document.c	2016-10-19 05:41:07.000000000 -0400
+++ php-5.3.29/ext/dom/document.c	2016-10-19 06:02:49.792006093 -0400
@@ -1845,7 +1845,7 @@ PHP_FUNCTION(dom_document_savexml)
 		if (options & LIBXML_SAVE_NOEMPTYTAG) {
 			xmlSaveNoEmptyTags = saveempty;
 		}
-		if (!size) {
+		if (!size || !mem) {
 			RETURN_FALSE;
 		}
 		RETVAL_STRINGL(mem, size, 1);
@@ -2352,7 +2352,7 @@ PHP_FUNCTION(dom_document_save_html)
 #else
 		htmlDocDumpMemory(docp, &mem, &size);
 #endif
-		if (!size) {
+		if (!size || !mem) {
 			RETVAL_FALSE;
 		} else {
 			RETVAL_STRINGL((const char*) mem, size, 1);
