diff -apruN php-5.3.29.orig/ext/mbstring/mbstring.c php-5.3.29/ext/mbstring/mbstring.c
--- php-5.3.29.orig/ext/mbstring/mbstring.c	2014-08-13 15:22:50.000000000 -0400
+++ php-5.3.29/ext/mbstring/mbstring.c	2016-10-20 05:24:47.486932566 -0400
@@ -212,7 +212,7 @@ static const struct mb_overload_def mb_o
 	{MB_OVERLOAD_REGEX, "split", "mb_split", "mb_orig_split"},
 #endif
 	{0, NULL, NULL, NULL}
-}; 
+};
 /* }}} */
 
 /* {{{ arginfo */
@@ -611,7 +611,7 @@ static void *_php_mb_allocators_calloc(u
 static void _php_mb_allocators_free(void *ptr)
 {
 	efree(ptr);
-} 
+}
 
 static void *_php_mb_allocators_pmalloc(unsigned int sz)
 {
@@ -626,7 +626,7 @@ static void *_php_mb_allocators_prealloc
 static void _php_mb_allocators_pfree(void *ptr)
 {
 	pefree(ptr, 1);
-} 
+}
 
 static mbfl_allocators _php_mb_allocators = {
 	_php_mb_allocators_malloc,
@@ -649,7 +649,7 @@ static sapi_post_entry mbstr_post_entrie
 
 /* {{{ static int php_mb_parse_encoding_list()
  *  Return 0 if input contains any illegal encoding, otherwise 1.
- *  Even if any illegal encoding is detected the result may contain a list 
+ *  Even if any illegal encoding is detected the result may contain a list
  *  of parsed encodings.
  */
 static int
@@ -774,13 +774,13 @@ php_mb_parse_encoding_list(const char *v
 
 /* {{{ MBSTRING_API php_mb_check_encoding_list */
 MBSTRING_API int php_mb_check_encoding_list(const char *encoding_list TSRMLS_DC) {
-	return php_mb_parse_encoding_list(encoding_list, strlen(encoding_list), NULL, NULL, 0 TSRMLS_CC);	
+	return php_mb_parse_encoding_list(encoding_list, strlen(encoding_list), NULL, NULL, 0 TSRMLS_CC);
 }
 /* }}} */
 
 /* {{{ static int php_mb_parse_encoding_array()
  *  Return 0 if input contains any illegal encoding, otherwise 1.
- *  Even if any illegal encoding is detected the result may contain a list 
+ *  Even if any illegal encoding is detected the result may contain a list
  *  of parsed encodings.
  */
 static int
@@ -817,7 +817,7 @@ php_mb_parse_encoding_array(zval *array,
 				if (strcasecmp(Z_STRVAL_PP(hash_entry), "auto") == 0) {
 					if (!bauto) {
 						bauto = 1;
-						l = identify_list_size; 
+						l = identify_list_size;
 						src = identify_list;
 						while (l > 0) {
 							*entry++ = *src++;
@@ -1053,7 +1053,7 @@ int _php_mb_ini_mbstring_internal_encodi
 	enum mbfl_no_encoding no_encoding;
   	const char *enc_name = NULL;
   	uint enc_name_len = 0;
-   
+
   	no_encoding = new_value ? mbfl_name2no_encoding(new_value):
   				mbfl_no_encoding_invalid;
 	if (no_encoding != mbfl_no_encoding_invalid) {
@@ -1266,13 +1266,13 @@ PHP_INI_BEGIN()
 	PHP_INI_ENTRY("mbstring.script_encoding", NULL, PHP_INI_ALL, OnUpdate_mbstring_script_encoding)
 #endif /* ZEND_MULTIBYTE */
 	PHP_INI_ENTRY("mbstring.substitute_character", NULL, PHP_INI_ALL, OnUpdate_mbstring_substitute_character)
-	STD_PHP_INI_ENTRY("mbstring.func_overload", "0", 
+	STD_PHP_INI_ENTRY("mbstring.func_overload", "0",
 	PHP_INI_SYSTEM, OnUpdateLong, func_overload, zend_mbstring_globals, mbstring_globals)
 
 	STD_PHP_INI_BOOLEAN("mbstring.encoding_translation", "0",
 		PHP_INI_SYSTEM | PHP_INI_PERDIR,
-		OnUpdate_mbstring_encoding_translation, 
-		encoding_translation, zend_mbstring_globals, mbstring_globals)					 
+		OnUpdate_mbstring_encoding_translation,
+		encoding_translation, zend_mbstring_globals, mbstring_globals)
 	PHP_INI_ENTRY("mbstring.http_output_conv_mimetypes",
 		"^(text/|application/xhtml\\+xml)",
 		PHP_INI_ALL,
@@ -1383,7 +1383,7 @@ PHP_MINIT_FUNCTION(mbstring)
 PHP_MSHUTDOWN_FUNCTION(mbstring)
 {
 	UNREGISTER_INI_ENTRIES();
-	
+
 #if HAVE_MBREGEX
 	PHP_MSHUTDOWN(mb_regex) (INIT_FUNC_ARGS_PASSTHRU);
 #endif
@@ -1427,21 +1427,21 @@ PHP_RINIT_FUNCTION(mbstring)
  	/* override original function. */
 	if (MBSTRG(func_overload)){
 		p = &(mb_ovld[0]);
-		
+
 		while (p->type > 0) {
-			if ((MBSTRG(func_overload) & p->type) == p->type && 
+			if ((MBSTRG(func_overload) & p->type) == p->type &&
 				zend_hash_find(EG(function_table), p->save_func,
 					strlen(p->save_func)+1, (void **)&orig) != SUCCESS) {
 
 				zend_hash_find(EG(function_table), p->ovld_func, strlen(p->ovld_func)+1 , (void **)&func);
-				
+
 				if (zend_hash_find(EG(function_table), p->orig_func, strlen(p->orig_func)+1, (void **)&orig) != SUCCESS) {
 					php_error_docref("ref.mbstring" TSRMLS_CC, E_WARNING, "mbstring couldn't find function %s.", p->orig_func);
 					return FAILURE;
 				} else {
 					zend_hash_add(EG(function_table), p->save_func, strlen(p->save_func)+1, orig, sizeof(zend_function), NULL);
 
-					if (zend_hash_update(EG(function_table), p->orig_func, strlen(p->orig_func)+1, func, sizeof(zend_function), 
+					if (zend_hash_update(EG(function_table), p->orig_func, strlen(p->orig_func)+1, func, sizeof(zend_function),
 						NULL) == FAILURE) {
 						php_error_docref("ref.mbstring" TSRMLS_CC, E_WARNING, "mbstring couldn't replace function %s.", p->orig_func);
 						return FAILURE;
@@ -1491,10 +1491,10 @@ PHP_RSHUTDOWN_FUNCTION(mbstring)
 	if (MBSTRG(func_overload)){
 		p = &(mb_ovld[0]);
 		while (p->type > 0) {
-			if ((MBSTRG(func_overload) & p->type) == p->type && 
+			if ((MBSTRG(func_overload) & p->type) == p->type &&
 				zend_hash_find(EG(function_table), p->save_func,
 							   strlen(p->save_func)+1, (void **)&orig) == SUCCESS) {
-				
+
 				zend_hash_update(EG(function_table), p->orig_func, strlen(p->orig_func)+1, orig, sizeof(zend_function), NULL);
 				zend_hash_del(EG(function_table), p->save_func, strlen(p->save_func)+1);
 			}
@@ -1516,7 +1516,7 @@ PHP_MINFO_FUNCTION(mbstring)
 	php_info_print_table_start();
 	php_info_print_table_row(2, "Multibyte Support", "enabled");
 	php_info_print_table_row(2, "Multibyte string engine", "libmbfl");
-	php_info_print_table_row(2, "HTTP input encoding translation", MBSTRG(encoding_translation) ? "enabled": "disabled");	
+	php_info_print_table_row(2, "HTTP input encoding translation", MBSTRG(encoding_translation) ? "enabled": "disabled");
 	php_info_print_table_end();
 
 	php_info_print_table_start();
@@ -1655,7 +1655,7 @@ PHP_FUNCTION(mb_http_input)
 						temp = list;
 						spprintf(&list, 0, "%s,%s", temp, name);
 						efree(temp);
-						if (!list) { 
+						if (!list) {
 							break;
 						}
 					} else {
@@ -1895,13 +1895,13 @@ PHP_FUNCTION(mb_parse_str)
 	encstr = estrndup(encstr, encstr_len);
 
 	info.data_type              = PARSE_STRING;
-	info.separator              = PG(arg_separator).input; 
+	info.separator              = PG(arg_separator).input;
 	info.force_register_globals = (track_vars_array == NULL);
 	info.report_errors          = 1;
 	info.to_encoding            = MBSTRG(current_internal_encoding);
 	info.to_language            = MBSTRG(language);
 	info.from_encodings         = MBSTRG(http_input_list);
-	info.num_from_encodings     = MBSTRG(http_input_list_size); 
+	info.num_from_encodings     = MBSTRG(http_input_list_size);
 	info.from_language          = MBSTRG(language);
 
 	detected = _php_mb_encoding_handler_ex(&info, track_vars_array, encstr TSRMLS_CC);
@@ -1967,7 +1967,7 @@ PHP_FUNCTION(mb_output_handler)
  		if (SG(sapi_headers).send_default_content_type || send_text_mimetype) {
 			charset = mbfl_no2preferred_mime_name(encoding);
 			if (charset) {
-				len = spprintf( &p, 0, "Content-Type: %s; charset=%s",  mimetype, charset ); 
+				len = spprintf( &p, 0, "Content-Type: %s; charset=%s",  mimetype, charset );
 				if (sapi_add_header(p, len, 0) != FAILURE) {
 					SG(sapi_headers).send_default_content_type = 0;
 				}
@@ -1990,7 +1990,7 @@ PHP_FUNCTION(mb_output_handler)
  	/* mode */
  	mbfl_buffer_converter_illegal_mode(MBSTRG(outconv), MBSTRG(current_filter_illegal_mode));
  	mbfl_buffer_converter_illegal_substchar(MBSTRG(outconv), MBSTRG(current_filter_illegal_substchar));
- 
+
  	/* feed the string */
  	mbfl_string_init(&string);
  	string.no_language = MBSTRG(language);
@@ -2000,11 +2000,11 @@ PHP_FUNCTION(mb_output_handler)
  	mbfl_buffer_converter_feed(MBSTRG(outconv), &string);
  	if (last_feed) {
  		mbfl_buffer_converter_flush(MBSTRG(outconv));
-	} 
+	}
  	/* get the converter output, and return it */
  	mbfl_buffer_converter_result(MBSTRG(outconv), &result);
  	RETVAL_STRINGL((char *)result.val, result.len, 0);		/* the string is already strdup()'ed */
- 
+
  	/* delete the converter if it is the last feed. */
  	if (last_feed) {
 		MBSTRG(illegalchars) += mbfl_buffer_illegalchars(MBSTRG(outconv));
@@ -2058,7 +2058,7 @@ PHP_FUNCTION(mb_strpos)
 	mbfl_string haystack, needle;
 	char *enc_name = NULL;
 	int enc_name_len;
-	
+
 	mbfl_string_init(&haystack);
 	mbfl_string_init(&needle);
 	haystack.no_language = MBSTRG(language);
@@ -2946,7 +2946,7 @@ PHP_FUNCTION(mb_convert_case)
 
 	if (newstr) {
 		RETVAL_STRINGL(newstr, ret_len, 0);
-	}	
+	}
 }
 /* }}} */
 
@@ -3168,7 +3168,7 @@ PHP_FUNCTION(mb_encode_mimeheader)
 	mbfl_string_init(&result);
 	ret = mbfl_mime_header_encode(&string, &result, charset, transenc, linefeed, indent);
 	if (ret != NULL) {
-		RETVAL_STRINGL((char *)ret->val, ret->len, 0)	/* the string is already strdup()'ed */
+		RETVAL_STRINGL_CHECK((char *)ret->val, ret->len, 0);	/* the string is already strdup()'ed */
 	} else {
 		RETVAL_FALSE;
 	}
@@ -3192,7 +3192,7 @@ PHP_FUNCTION(mb_decode_mimeheader)
 	mbfl_string_init(&result);
 	ret = mbfl_mime_header_decode(&string, &result, MBSTRG(current_internal_encoding));
 	if (ret != NULL) {
-		RETVAL_STRINGL((char *)ret->val, ret->len, 0)	/* the string is already strdup()'ed */
+		RETVAL_STRINGL_CHECK((char *)ret->val, ret->len, 0);	/* the string is already strdup()'ed */
 	} else {
 		RETVAL_FALSE;
 	}
@@ -3295,7 +3295,7 @@ PHP_FUNCTION(mb_convert_kana)
 
 	ret = mbfl_ja_jp_hantozen(&string, &result, opt);
 	if (ret != NULL) {
-		RETVAL_STRINGL((char *)ret->val, ret->len, 0);		/* the string is already strdup()'ed */
+		RETVAL_STRINGL_CHECK((char *)ret->val, ret->len, 0);		/* the string is already strdup()'ed */
 	} else {
 		RETVAL_FALSE;
 	}
@@ -3317,8 +3317,8 @@ PHP_FUNCTION(mb_convert_variables)
 	int n, to_enc_len, argc, stack_level, stack_max, elistsz;
 	enum mbfl_no_encoding *elist;
 	char *name, *to_enc;
-	void *ptmp;	
-	
+	void *ptmp;
+
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "sZ+", &to_enc, &to_enc_len, &zfrom_enc, &args, &argc) == FAILURE) {
 		return;
 	}
@@ -3584,7 +3584,7 @@ php_mb_numericentity_exec(INTERNAL_FUNCT
 
 	ret = mbfl_html_numeric_entity(&string, &result, convmap, mapsize, type);
 	if (ret != NULL) {
-		RETVAL_STRINGL((char *)ret->val, ret->len, 0);
+		RETVAL_STRINGL_CHECK((char *)ret->val, ret->len, 0);
 	} else {
 		RETVAL_FALSE;
 	}
@@ -3672,12 +3672,12 @@ static int _php_mbstr_parse_mail_headers
 
 	/*
 	 *             C o n t e n t - T y p e :   t e x t / h t m l \r\n
-	 *             ^ ^^^^^^^^^^^^^^^^^^^^^ ^^^ ^^^^^^^^^^^^^^^^^ ^^^^ 
-	 *      state  0            1           2          3          
+	 *             ^ ^^^^^^^^^^^^^^^^^^^^^ ^^^ ^^^^^^^^^^^^^^^^^ ^^^^
+	 *      state  0            1           2          3
 	 *
 	 *             C o n t e n t - T y p e :   t e x t / h t m l \r\n
-	 *             ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^ 
-	 * crlf_state -1                       0                     1 -1  
+	 *             ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^^^^
+	 * crlf_state -1                       0                     1 -1
 	 *
 	 */
 
@@ -3743,7 +3743,7 @@ static int _php_mbstr_parse_mail_headers
 						token.a = 0;
 						state = 1;
 						break;
-					
+
 					case 2:
 						if (crlf_state != -1) {
 							token.c = (char *)ps;
@@ -3862,12 +3862,12 @@ PHP_FUNCTION(mb_send_mail)
 	smart_str *s;
 	extern void mbfl_memory_device_unput(mbfl_memory_device *device);
 	char *pp, *ee;
-    
+
 	if (PG(safe_mode) && (ZEND_NUM_ARGS() == 5)) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "SAFE MODE Restriction in effect.  The fifth parameter is disabled in SAFE MODE.");
 		RETURN_FALSE;
 	}
-    
+
 	/* initialize */
 	mbfl_memory_device_init(&device, 0, 0);
 	mbfl_string_init(&orig_str);
@@ -3925,14 +3925,14 @@ PHP_FUNCTION(mb_send_mail)
 				if ((param_name = php_strtok_r(p, "= ", &tmp)) != NULL) {
 					if (strcasecmp(param_name, "charset") == 0) {
 						enum mbfl_no_encoding _tran_cs = tran_cs;
-						
+
 						charset = php_strtok_r(NULL, "= \"", &tmp);
 						if (charset != NULL) {
 							_tran_cs = mbfl_name2no_encoding(charset);
 						}
 
 						if (_tran_cs == mbfl_no_encoding_invalid) {
-							php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unsupported charset \"%s\" - will be regarded as ascii", charset); 
+							php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unsupported charset \"%s\" - will be regarded as ascii", charset);
 							_tran_cs = mbfl_no_encoding_ascii;
 						}
 						tran_cs = _tran_cs;
@@ -3957,7 +3957,7 @@ PHP_FUNCTION(mb_send_mail)
 				break;
 
 			default:
-				php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unsupported transfer encoding \"%s\" - will be regarded as 8bit", s->c); 
+				php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unsupported transfer encoding \"%s\" - will be regarded as 8bit", s->c);
 				body_enc =	mbfl_no_encoding_8bit;
 				break;
 		}
@@ -4090,7 +4090,7 @@ PHP_FUNCTION(mb_send_mail)
 		extra_cmd = php_escape_shell_cmd(force_extra_parameters);
 	} else if (extra_cmd) {
 		extra_cmd = php_escape_shell_cmd(extra_cmd);
-	} 
+	}
 
 	if (!err && php_mail(to_r, subject, message, headers, extra_cmd TSRMLS_CC)) {
 		RETVAL_TRUE;
@@ -4191,7 +4191,7 @@ PHP_FUNCTION(mb_get_info)
 		}
 		if ((name = (char *)mbfl_no_language2name(MBSTRG(language))) != NULL) {
 			add_assoc_string(return_value, "language", name, 1);
-		}		
+		}
 		n = MBSTRG(current_detect_order_list_size);
 		entry = MBSTRG(current_detect_order_list);
 		if(n > 0) {
@@ -4239,15 +4239,15 @@ PHP_FUNCTION(mb_get_info)
 	} else if (!strcasecmp("internal_encoding", typ)) {
 		if ((name = (char *)mbfl_no_encoding2name(MBSTRG(current_internal_encoding))) != NULL) {
 			RETVAL_STRING(name, 1);
-		}		
+		}
 	} else if (!strcasecmp("http_input", typ)) {
 		if ((name = (char *)mbfl_no_encoding2name(MBSTRG(http_input_identify))) != NULL) {
 			RETVAL_STRING(name, 1);
-		}		
+		}
 	} else if (!strcasecmp("http_output", typ)) {
 		if ((name = (char *)mbfl_no_encoding2name(MBSTRG(current_http_output_encoding))) != NULL) {
 			RETVAL_STRING(name, 1);
-		}		
+		}
 	} else if (!strcasecmp("http_output_conv_mimetypes", typ)) {
 		if ((name = (char *)zend_ini_string("mbstring.http_output_conv_mimetypes", sizeof("mbstring.http_output_conv_mimetypes"), 0)) != NULL) {
 			RETVAL_STRING(name, 1);
@@ -4290,7 +4290,7 @@ PHP_FUNCTION(mb_get_info)
 	} else if (!strcasecmp("language", typ)) {
 		if ((name = (char *)mbfl_no_language2name(MBSTRG(language))) != NULL) {
 			RETVAL_STRING(name, 1);
-		}		
+		}
 	} else if (!strcasecmp("detect_order", typ)) {
 		n = MBSTRG(current_detect_order_list_size);
 		entry = MBSTRG(current_detect_order_list);
@@ -4373,15 +4373,15 @@ PHP_FUNCTION(mb_check_encoding)
 			RETURN_FALSE;
 		}
 	}
-	
+
 	convd = mbfl_buffer_converter_new(no_encoding, no_encoding, 0);
 	if (convd == NULL) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Unable to create converter");
 		RETURN_FALSE;
-	}	
+	}
 	mbfl_buffer_converter_illegal_mode(convd, MBFL_OUTPUTFILTER_ILLEGAL_MODE_NONE);
-	mbfl_buffer_converter_illegal_substchar(convd, 0);	
-	
+	mbfl_buffer_converter_illegal_substchar(convd, 0);
+
 	/* initialize string */
 	mbfl_string_init_set(&string, mbfl_no_language_neutral, no_encoding);
 	mbfl_string_init(&result);
@@ -4403,7 +4403,7 @@ PHP_FUNCTION(mb_check_encoding)
 /* }}} */
 
 /* {{{ MBSTRING_API int php_mb_encoding_translation() */
-MBSTRING_API int php_mb_encoding_translation(TSRMLS_D) 
+MBSTRING_API int php_mb_encoding_translation(TSRMLS_D)
 {
 	return MBSTRG(encoding_translation);
 }
@@ -4521,7 +4521,7 @@ MBSTRING_API int php_mb_gpc_encoding_con
 		}
 	} else {
 		to_encoding = MBSTRG(current_internal_encoding);
-	}	
+	}
 	if (encoding_from) {
 		/* old encoding */
 		from_encoding = mbfl_name2no_encoding(encoding_from);
@@ -4553,7 +4553,7 @@ MBSTRING_API int php_mb_gpc_encoding_con
 		}
 		mbfl_buffer_converter_illegal_mode(convd, MBSTRG(current_filter_illegal_mode));
 		mbfl_buffer_converter_illegal_substchar(convd, MBSTRG(current_filter_illegal_substchar));
-		
+
 		/* do it */
 		ret = mbfl_buffer_converter_feed_result(convd, &string, &result);
 		if (ret != NULL) {
@@ -4561,7 +4561,7 @@ MBSTRING_API int php_mb_gpc_encoding_con
 			str[i] = (char *)ret->val;
 			len[i] = (int)ret->len;
 		}
-		
+
 		MBSTRG(illegalchars) += mbfl_buffer_illegalchars(convd);
 		mbfl_buffer_converter_delete(convd);
 	}
@@ -4577,18 +4577,18 @@ MBSTRING_API int php_mb_gpc_encoding_det
 	mbfl_string string;
 	enum mbfl_no_encoding *elist;
 	enum mbfl_no_encoding encoding = mbfl_no_encoding_invalid;
-	mbfl_encoding_detector *identd = NULL; 
+	mbfl_encoding_detector *identd = NULL;
 
 	int size;
 	enum mbfl_no_encoding *list;
 
-	if (MBSTRG(http_input_list_size) == 1 && 
+	if (MBSTRG(http_input_list_size) == 1 &&
 		MBSTRG(http_input_list)[0] == mbfl_no_encoding_pass) {
 		MBSTRG(http_input_identify) = mbfl_no_encoding_pass;
 		return SUCCESS;
 	}
 
-	if (MBSTRG(http_input_list_size) == 1 && 
+	if (MBSTRG(http_input_list_size) == 1 &&
 		MBSTRG(http_input_list)[0] != mbfl_no_encoding_auto &&
 		mbfl_no_encoding2name(MBSTRG(http_input_list)[0]) != NULL) {
 		MBSTRG(http_input_identify) = MBSTRG(http_input_list)[0];
@@ -4600,7 +4600,7 @@ MBSTRING_API int php_mb_gpc_encoding_det
 		list = NULL;
 		size = 0;
 		php_mb_parse_encoding_list(arg_list, strlen(arg_list), &list, &size, 0 TSRMLS_CC);
-		
+
 		if (size > 0 && list != NULL) {
 			elist = list;
 		} else {
@@ -4695,7 +4695,7 @@ MBSTRING_API int php_mb_stripos(int mode
 
  		{
  			int haystack_char_len = mbfl_strlen(&haystack);
- 
+
  			if (mode) {
  				if ((offset > 0 && offset > haystack_char_len) ||
  					(offset < 0 && -offset > haystack_char_len)) {
@@ -4827,7 +4827,7 @@ static int php_mb_encoding_converter(uns
 	to_encoding = mbfl_name2no_encoding(encoding_to);
 	if (to_encoding == mbfl_no_encoding_invalid) {
 		return -1;
-	}	
+	}
 	/* old encoding */
 	from_encoding = mbfl_name2no_encoding(encoding_from);
 	if (from_encoding == mbfl_no_encoding_invalid) {
